lib/db/backend/badger_backend.go

```diff
@@ -0,0 +1,377 @@
+	select {
+	case <-b.closed:
+		return nil, errClosed{}
+	default:
+	}
+
+	txn := b.bdb.NewTransaction(false)
+	it := badgerRangeIterator(txn, first, last)
+	it.releaseFn = func() {
+		txn.Discard()
+	}
+	return it, nil
+}
+
+func (b *badgerBackend) Put(key, val []byte) error {
+	select {
+	case <-b.closed:
+		return errClosed{}
+	default:
+	}
+
+	txn := b.bdb.NewTransaction(true)
+	if err := txn.Set(key, val); err != nil {
+		txn.Discard()
+		return wrapBadgerErr(err)
+	}
+	return wrapBadgerErr(txn.Commit())
+}
+
+func (b *badgerBackend) Delete(key []byte) error {
+	select {
+	case <-b.closed:
+		return errClosed{}
+	default:
+	}
+
+	txn := b.bdb.NewTransaction(true)
+	if err := txn.Delete(key); err != nil {
+		txn.Discard()
+		return wrapBadgerErr(err)
+	}
+	return wrapBadgerErr(txn.Commit())
+}
+
+func (b *badgerBackend) Compact() error {
+	select {
+	case <-b.closed:
+		return errClosed{}
+	default:
+	}
+
+	// XXX: check if this is appropriate or if we also need Flatten and
+	// whatnot. Also it apparently returns an error if the GC didn't result
+	// in anything being freed...
+	_ = b.bdb.RunValueLogGC(0.5)
+	return nil
+}
+
+// badgerSnapshot implements backend.ReadTransaction
+type badgerSnapshot struct {
+	txn *badger.Txn
+	rel *releaser
+}
+
+func (l badgerSnapshot) Get(key []byte) ([]byte, error) {
+	item, err := l.txn.Get(key)
+	if err != nil {
+		return nil, wrapBadgerErr(err)
+	}
+	val, err := item.ValueCopy(nil)
+	if err != nil {
+		return nil, wrapBadgerErr(err)
+	}
+	return val, nil
+}
+
+func (l badgerSnapshot) NewPrefixIterator(prefix []byte) (Iterator, error) {
+	return badgerPrefixIterator(l.txn, prefix), nil
+}
+
+func (l badgerSnapshot) NewRangeIterator(first, last []byte) (Iterator, error) {
+	return badgerRangeIterator(l.txn, first, last), nil
+}
+
+func (l badgerSnapshot) Release() {
+	defer l.rel.Release()
+	l.txn.Discard()
+}
+
+// badgerTransaction implements backend.WriteTransaction using a batch (not
+// an actual badger transaction)
+type badgerTransaction struct {
+	badgerSnapshot
+	txn  *badger.Txn
+	bdb  *badger.DB
+	rel  *releaser
+	size int
+}
+
+func (t *badgerTransaction) Delete(key []byte) error {
+	t.size += len(key)
+	kc := make([]byte, len(key))
+	copy(kc, key)
+	return t.transactionRetried(func(txn *badger.Txn) error {
+		return txn.Delete(kc)
+	})
+}
+
+func (t *badgerTransaction) Put(key, val []byte) error {
+	t.size += len(key) + len(val)
+	kc := make([]byte, len(key))
+	copy(kc, key)
+	vc := make([]byte, len(val))
+	copy(vc, val)
+	return t.transactionRetried(func(txn *badger.Txn) error {
+		return txn.Set(kc, vc)
+	})
+}
+
+// transactionRetried performs the given operation in the current
+// transaction, with commit and retry if Badger says the transaction has
+// grown too large.
+func (t *badgerTransaction) transactionRetried(fn func(*badger.Txn) error) error {
+	if err := fn(t.txn); err == badger.ErrTxnTooBig {
+		if err := t.txn.Commit(); err != nil {
```

Comment:
This sort of kills batch atomicity. Can we guide the batch sizes if we know what the max txn size is instead of splitting it?

Reply:
Yeah this is a wart. I don't think it kicks in until the transaction is many, many megabytes large (way after our normal batch sizes), but it's not documented at all. I'll do some testing.

Reply:
This currently checks transaction size on checkpointing only though - we should also do it on every write/del with a higher limit like on leveldb.

Reply:
Well, in LevelDB this is a thing because the "transaction" is an in-memory batch and we don't want to let it grow too much. In Badger it's an actual transaction on disk as far as I can tell, so we don't actually have something in memory that we must limit.

If we call Checkpoint() we're pretty much guaranteed to never hit the transaction-too-large-auto-commit and if we don't, then I don't think it matters?